### ISSUE TITLE
Contributte/datagrid false positive

### ISSUE THREAD

**Issue by jakubvojacek** [efabrica-team/phpstan-latte#98]
Hello

we're heavily using datagrid and it emmits 3-4 false positives (at least i think they're false positive) per control

```
 ------ -------------------------------------------------------------------------------------------------------------------- 
  Line   FrontModule/Components/Config/VendorGrid.latte rendered from Motv\FrontModule\Components\Config\VendorGrid::render  
 ------ -------------------------------------------------------------------------------------------------------------------- 
  1      Cannot call method render() on void.                                                                                
  1      Instanceof between void and Nette\Application\UI\Renderable will always evaluate to false.                          
  1      Only booleans are allowed in an if condition, bool|null given.                                                      
  1      Result of method Nette\ComponentModel\Container::getComponent() (void) is used.                                     
 ------ -------------------------------------------------------------------------------------------------------------------- 
```

## vendorGrid.latte
```latte
{control vendorGrid}
```

similarily with just snippet

```
 ------ -------------------------------------------------------------------------------------------------------------------------------- 
  Line   FrontModule/Components/Config/VendorAvatarGrid.latte rendered from Motv\FrontModule\Components\Config\VendorAvatarGrid::render  
 ------ -------------------------------------------------------------------------------------------------------------------------------- 
  1      Only booleans are allowed in an if condition, bool|null given.                                                                  
  2      Cannot call method render() on void.                                                                                            
  2      Instanceof between void and Nette\Application\UI\Renderable will always evaluate to false.                                      
  2      Result of method Nette\ComponentModel\Container::getComponent() (void) is used.                                                 
 ------ -------------------------------------------------------------------------------------------------------------------------------- 
```

##VendorAvatarGrid.latte
```
{snippet}
	{control vendorAvatarGrid}
{/snippet}
```

Should this be added into the ignored errors (https://github.com/efabrica-team/phpstan-latte/blob/main/phpstan.neon)?  Or is there something truly wrong with my latte code?

Thanks

**Comment by MartinMystikJonas** [comment]
Could you sent us content od compiled template code from tmp/phpstan-latte?

**Comment by MartinMystikJonas** [comment]
And code of createComponentAvatarGrid (including doc comment)

**Comment by lulco** [comment]
It seems createComponentAvatarGrid has void return type. But it shouldn't work in app or is it?

**Comment by MartinMystikJonas** [comment]
Maybe wrong `@return` annotation? But it should be reported by PHPStan that void method returns something. Only explanation I can think of is that createComponentAvatarGrid is outside analysed project and have `@return void`.

**Comment by jakubvojacek** [comment]
Yes, it actually has `void`return type, this is how it was described in datagrid docs (https://contributte.org/packages/contributte/datagrid/#features at the bottom of page) and it worked, never really given it much thought and components are just copy paste from existing and changing columns, renderes etc..

After adding the return type, it now complaints about:

```
 ------ -------------------------------------------------------------------------------------------------------------------------------------- 
  Line   FrontModule/Components/Config/VendorAvatarGrid.latte rendered from Motv\FrontModule\Components\Config\VendorAvatarGrid::render        
 ------ -------------------------------------------------------------------------------------------------------------------------------------- 
  1      Only booleans are allowed in an if condition, bool|null given.                                                                        
  2      Instanceof between Motv\FrontModule\Components\Common\CsmsDataGrid and Nette\Application\UI\Renderable will always evaluate to true.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------- 
```

I would share the generated code from tmp folder but it no longer exists - I remember that previous versions created tmp folder in root directory but its no longer happening.

**Comment by MartinMystikJonas** [comment]
Default now is in system temp (usually /tmp/phpstan-latte) or you can set different directory this way:
```
parameters:
    latte:
        tmpDir: /path/to/tmp
```

Note: PHPStan currently cannot resolve relative paths in custom config values. It is better to set absolute path. Relative paths would be relative to cwd.

**Comment by MartinMystikJonas** [comment]
My my guess is taht both these errors are coused by phpstan/strict-rules applied to generated code that is generated for less specific types that we know when anylysing it. So ti could be ignored. We should probably add second one to default ignore patterns. Not sure about the first one - we would need that compiled code to know for sure.

**Comment by jakubvojacek** [comment]
> My my guess is taht both these errors are coused by phpstan/strict-rules

I do use this extension, yes

```php
<?php

declare (strict_types=1);
use Latte\Runtime as LR;
/** 
* Motv\FrontModule\Components\Config\VendorAvatarGrid::render
* @property PHPStanLatteTemplate_31e42b883c180517be5295c9f191f372_global $global
*/
final class PHPStanLatteTemplate_31e42b883c180517be5295c9f191f372 extends Latte\Runtime\Template
{
    public const Blocks = ['snippet' => ['' => 'block1']];
    public function main(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var array<string> $ʟ_tag */
        /** @var array $ʟ_if */
        /** @var array $ʟ_ifc */
        /** @var array $ʟ_try */
        /** @var array $ʟ_loc */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var Motv\FrontModule\Components\Config\VendorAvatarGrid $presenter */
        /** @var Motv\FrontModule\Components\Config\VendorAvatarGrid $control */
        unset($ʟ_args);
        if ($this->global->snippetDriver?->renderSnippets($this->blocks[self::LayerSnippet], $this->params)) {
            return;
        }
        echo '<div id="', htmlspecialchars($this->global->snippetDriver->getHtmlId('')), '">';
        /* line 1 */
        $this->renderBlock('', [], null, 'snippet');
        echo '</div>
';
    }
    /** {snippet} on line 1 */
    public function block1(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var array<string> $ʟ_tag */
        /** @var array $ʟ_if */
        /** @var array $ʟ_ifc */
        /** @var array $ʟ_try */
        /** @var array $ʟ_loc */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var Motv\FrontModule\Components\Config\VendorAvatarGrid $presenter */
        /** @var Motv\FrontModule\Components\Config\VendorAvatarGrid $control */
        unset($ʟ_args);
        /* line 1 */
        $this->global->snippetDriver->enter('', 'static');
        try {
            echo "\n";
            /* line 2 */
            /** @var Motv\FrontModule\Components\Common\CsmsDataGrid $ʟ_tmp */
            $ʟ_tmp = $this->global->uiControl->getComponent('vendorAvatarGrid');
            if ($ʟ_tmp instanceof Nette\Application\UI\Renderable) {
                $ʟ_tmp->redrawControl(null, false);
            }
            $ʟ_tmp->render();
        } finally {
            $this->global->snippetDriver->leave();
        }
    }
}
/**
* @property-read array $coreParentFinder
* @property-read Motv\FrontModule\Components\Config\VendorAvatarGrid $uiControl
* @property-read Motv\FrontModule\Components\Config\VendorAvatarGrid $uiPresenter
* @property-read Nette\Bridges\ApplicationLatte\SnippetDriver|null $snippetDriver
* @property-read string|null $uiNonce
* @property-read Kdyby\Redis\RedisStorage $cacheStorage
*/
class PHPStanLatteTemplate_31e42b883c180517be5295c9f191f372_global extends \stdClass
{
}
```

**Comment by MartinMystikJonas** [comment]
First error is caused by this
```php
if ($this->global->snippetDriver?->renderSnippets($this->blocks[self::LayerSnippet], $this->params)) {
```

Second one by
```php
 if ($ʟ_tmp instanceof Nette\Application\UI\Renderable) {
```

Both can be ignored.

@lulco I would add them to default ignores (phpstan/strict-rules are quite common) in this form:

```
  '#Only booleans are allowed in an if condition, bool|null given.#'                                                              
  '#Instanceof between .* and Nette\\\\Application\\\\UI\\\\Renderable will always evaluate to true\.#'
```

Ok?

**Comment by lulco** [comment]
Yes please

**Comment by lulco** [comment]
> Default now is in system temp (usually /tmp/phpstan-latte) or you can set different directory this way:
> 
> ```
> parameters:
>     latte:
>         tmpDir: /path/to/tmp
> ```
> 
> Note: PHPStan currently cannot resolve relative paths in custom config values. It is better to set absolute path. Relative paths would be relative to cwd.

For me %currentWorkingDirectory%/tmp works

**Comment by lulco** [comment]
I don’t get it how that grid works :) if it is void, how can ->render() be called on it? :)

**Comment by MartinMystikJonas** [comment]
My guess is that it adds itself to parent in constructor

**Comment by MartinMystikJonas** [comment]
Should we analyse constructors of objects created in createdComponent if `$this` is passed? 🤔

**Comment by lulco** [comment]
It is done by collectors

**Comment by MartinMystikJonas** [comment]
Well not in this case. Worth checking it.

**Comment by MartinMystikJonas** [comment]
Oh of course it is caused by #85

**Comment by MartinMystikJonas** [comment]
@jakubvojacek Can you check current main - it should ignore that scrict rules errors generated by latte

**Comment by jakubvojacek** [comment]
yes, works perfectly 960 to 413 reported errors :) Most of which are the grid components with void return types, thanks alot :+1:

**Comment by MartinMystikJonas** [comment]
You can try add vendor/contributte/datagrid folder to analysedPaths. This should allow analyse of component creation datagrid does in constructor. It is workaround for #85 and might cause reporting of errors in contributte/datagrid itself (if there are any) but it could help.

**Comment by jakubvojacek** [comment]
Thanks, I added the return types, closing :+1:
